crates/red_knot_python_semantic/src/types.rs

```diff
@@ -1443,6 +1443,36 @@ impl<'db> Type<'db> {
         }
     }
 
+    pub(crate) fn is_none(self, db: &'db dyn Db) -> bool {
+        let Self::Instance(InstanceType { class }) = self else {
+            return false;
+        };
+
+        class.is_known(db, KnownClass::NoneType)
+    }
+
+    pub(crate) fn is_exception_class(self, db: &'db dyn Db) -> bool {
+        let subclass_of_base_exception = KnownClass::BaseException
+            .to_subclass_of(db)
+            .unwrap_or(Type::Unknown);
+
+        self.is_assignable_to(db, subclass_of_base_exception)
+    }
+
+    pub(crate) fn is_exception(self, db: &'db dyn Db) -> bool {
+        let instance_of_base_exception = KnownClass::BaseException.to_instance(db);
+
+        self.is_assignable_to(db, instance_of_base_exception)
+    }
+
+    pub(crate) fn is_raisable(self, db: &'db dyn Db) -> bool {
+        self.is_exception(db) || self.is_exception_class(db)
+    }
+
+    pub(crate) fn can_be_exception_cause(self, db: &'db dyn Db) -> bool {
+        self.is_raisable(db) || self.is_none(db)
+    }
```

Comment:
I'm interested in hearing other opinions. These methods feel too specific to me to have them on `Type`. That's why I'm leaning towards making them standalone functions. Unless we forsee using them in other places too (I want to avoid that `Type` ends up with a trillion very specific methods)

Reply:
I would like us to increase our use of well-named functions that describe some semantic of a type, rather than just inlining all of it into type inference, particularly when the same logic would be repeated in more than one location. I do think this makes type inference more readable. I think these methods are worth it if they have a sensible/clear name and have 2+ uses. If they have only one use, I agree that's questionable.

As for where they go, I would prefer to keep them on `Type`; I think that's better than a collection of standalone functions which operate on a single `Type`, both in usage ergonomics and in discoverability. I'm not worried about too many methods on `Type`, if we go with the heuristic that methods should generally have 2+ uses, and have a name that clearly represents something semantically meaningful.

Concretely in this case I think those heuristics would suggest keeping `is_exception_class` and `is_exception`, but not `is_raisable` or `can_be_exception_cause`. (I also think `is_none` is fine in general, but as Alex points out, maybe isn't the clearest option here -- though given that `None` is final and that will never change, it really doesn't matter in practice.)

Reply:
> Concretely in this case I think those heuristics would suggest keeping `is_exception_class` and `is_exception`, but not `is_raisable` or `can_be_exception_cause`.

`is_exception` is a one-liner, and I honestly find `.is_assignable_to(KnownClass::BaseException.to_instance(db))` more readable, because I don't have to lookup exactly what `is_exception()` is doing under the hood.

`is_exception_class` is a bit more complicated than `is_exception`, because `KnownClass::to_subclass_of()` returns `Option<Type>` rather than `Type`. That's probably a bad signature for `to_subclass_of()` to have, to be honest; I probably made a mistake there when I added it in https://github.com/astral-sh/ruff/commit/ab26d9cf9a5899406e1b4a3555c54e6d0be8ed96#diff-42314c006689490bbdfbeeb973de64046b3e069e3d88f67520aeba375f20e655R1910. Both `KnownClass::to_class_literal()` and `KnownClass::to_instance()` fall back to `Unknown` if they can't find the type being looked up rather than returning an `Option`; for consistency, `KnownClass::to_subclass_of()` should probably return `type[Unknown]` if it can't find the type in question, rather than returning an `Option`.

Reply:
Ok, I think it's fair to also add the heuristic "encapsulates some non-trivial logic" :) And I agree that making `to_subclass_of` just handle the fallback to `Unknown` internally is a good alternative to `is_exception_class`.